### ISSUE TITLE
Modifier key for snapping nodes on straight lines

### ISSUE THREAD

**Issue by famo** [openstreetmap/iD#6276]
Kinda hard to describe, so here is a picture
![Screenshot_20190430_231441](https://user-images.githubusercontent.com/692646/56994292-ec998680-6b9e-11e9-928b-2e91064b6f72.png)

So when dragging the orange node _and pressing an modifier key_, it should snap anywhere on the red dotted lines.

This would be really useful, when splitting up ways to add a `covered=yes` section and similar tasks.

**Comment by quincylvania** [comment]
I like this! I think we should just have the node snap automatically unless a modifier key is down, like how other snapping works in iD.

**Comment by manfredbrandl** [comment]
It should snap without modifier key preferable to the place where the red dotted lines cross.

**Comment by bhousel** [comment]
I kind of feel like this would be more annoying than useful 🤔 
I almost never want roads to snap this way when I'm adjusting nodes.   I definitely wouldn't want it to be the automatic behavior.

**Comment by BjornRasmussen** [comment]
I personally don't like this idea, since I would never use it.  If other people need it, a `shift` or `ctrl / cmnd` modifier should be required for the operation to behave in this way.

**Comment by bhousel** [comment]
I think I'd use it sometimes, but I'm mostly worried about making it a default behavior.   I'm worried it might do the wrong thing anytime a road is not straight.  Like squashing roundabouts or something:

<img width="412" alt="Screenshot 2019-05-01 14 50 53" src="https://user-images.githubusercontent.com/38784/57037455-40e74980-6c25-11e9-8da6-7775151dc046.png">

**Comment by sun-geo** [comment]
@famo 
> This would be really useful, when splitting up ways to add a `covered=yes` section and similar tasks.

Do you mean ways from building outlines (e.g. amenity=fuel roofs) and highways?
If so, actually for these cases I still wonder if there should be used shared nodes, because of the different layers ...

**Comment by famo** [comment]
To clarify I also don't need this as the default behavior, it should only be there on demand (e.g. by pressing an modifier key).

> Do you mean ways from building outlines (e.g. amenity=fuel roofs) and highways?
> If so, actually for these cases I still wonder if there should be used shared nodes, because of the different layers ...

I would agree on the layers, but even then you need to align covered with the non-covered way - which isn't possible ATM without workarounds.

**Comment by quincylvania** [comment]
Yeah I don't really agree anymore with my initial thought about making this default.

**Comment by cpmsmith** [comment]
I'd definitely like this; I've wished I had it a number of times. In particular when moving nodes along an existing straight way, it'd be nice for adjusting the position of the node without bending the way. I might take a crack at this in a few weeks, but in the meantime, I'll think through some details aloud:

* I agree it should be enabled with a modifier key, and coming from Photoshop et al, I'd expect it to be <kbd>shift</kbd>
* Trying to define the behaviour a little more formally, when dragging the node N:
  * Let A be the set of nodes which are adjacent to N in an existing way
  * Let L be the set of all possible lines drawn between nodes A
  * Let I be the set of intersections of lines L
  * If the nearest intersection of I is within snap distance (~10px?) of N, snap to it
  * Else if the nearest line of L is within snap distance of N, snap to the nearest point on that line
  * Else do not snap

It's starting to sound like it could potentially have a pretty bad algorithmic complexity, but I know optimizations for these things exist, probably even built into d3.

This also differs from the original proposal a bit, preferring the red dot here and adding more lines to which it could potentially snap, but I think that would make it a bit more flexible.

![56994292-ec998680-6b9e-11e9-928b-2e91064b6f72](https://user-images.githubusercontent.com/9437625/208904532-48a9c606-c8e3-45d1-a484-111beb57779d.png)
